### ISSUE TITLE
Do all attention layers perform equally good?

### ISSUE THREAD

**Issue by analyticsbot** [sagieppel/Classification-of-the-material-given-region-of-an-image-using-a-convolutional-neural-net-with-attent#2]
There is an attention layer in the main code, however there are other attention layers in the 'Unused_Alternative_Attention_Nets'.  Is there any metric to support which one of them performance the best?

**Comment by sagieppel** [comment]
This method  (addition first layer) seems to work the best but the difference in performance is not large.
See this doc for comparison:
https://arxiv.org/ftp/arxiv/papers/1812/1812.00291.pdf

**Comment by analyticsbot** [comment]
Got it. Are there any other new methods (since you wrote this) you could
suggest to improve the accuracy.

Especially in cases where there is quite an imbalance in the data. Using
focal loss is something one could use.

**Comment by sagieppel** [comment]
There are a billion methods:)
My opinion that diversifying the data and pretraining with other datasets, and picking the hard examples from the training set in higher frequency is more important than loss/net types.

**Comment by analyticsbot** [comment]
So oversample the hard to find examples + data augmentation?

If you have any papers that you can suggest, do share.

**Comment by sagieppel** [comment]
I don't know what is your dataset and how large it is, but in general pretraining with a general dataset like COCO. 
And co-training with other datasets related to your is from my experience the best approach.  If you have rare classes I suggest using the class balance in training.  I don't know about papers, it's more like common practices.
